File: day-2/go/second.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"strconv"
	"strings"
)

func second(scanner *bufio.Scanner) {
	hposition, depth, aim := 0, 0, 0

	for scanner.Scan() {
		split_text := strings.Split(scanner.Text(), " ")

		direction := split_text[0]
		number, err := strconv.Atoi(split_text[1])
		if err != nil {
			log.Fatal(err)
		}

		switch direction {
		case "up":
			aim -= number
		case "down":
			aim += number
		case "forward":
			hposition += number
			depth += aim * number
		}
	}

	fmt.Println("Second:", hposition*depth)

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}

File: day-2/nodejs/src/second.js
const parse = (coordinates, [depth, hposition, aim]) => {
    if (!coordinates.length)
        return [depth, hposition]

    const [command, number] = coordinates[0]

    switch (command) {
        case 'up':
            return parse(coordinates.slice(1), [depth, hposition, aim - number])

        case 'down':
            return parse(coordinates.slice(1), [depth, hposition, aim + number])

        default:
            return parse(coordinates.slice(1), [depth + (aim * number), hposition + number, aim])
    }
}

export const second = (coordinates) => {
    return parse(coordinates, [0, 0, 0]).reduce((a, b) => a * b, 1)
}